org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ChangeCalculator.java

```diff
@@ -0,0 +1,548 @@
+		ListRewrite typeParamsRewrite= fTargetMethodDeclarationASTRewrite.getListRewrite(fTargetMethodDeclaration, MethodDeclaration.TYPE_PARAMETERS_PROPERTY);
+		Javadoc javadoc= fTargetMethodDeclaration.getJavadoc();
+		List<String> methodParameterTypes= getMethodParameterTypes();
+		List<String> methodTypeParametersNames= getTypeParameterNames();
+
+		if (classTypeParameters.length != 0) {
+			for (int i= 0; i < classTypeParameters.length; i++) {
+				TypeParameter typeParameter= generateTypeParameter(classTypeParameters, i);
+
+				if (fFinalConditionsChecker.getStatus().hasError()) {
+					return;
+				}
+				//Check if method needs this TypeParameter (only if one or more methodParams are of this type OR method has instance usage OR an instance of parent class is used as methodParam)
+				String typeParamName= typeParameter.getName().getIdentifier();
+				String typeParamNameAsArray= typeParamName + "[]"; //$NON-NLS-1$
+				boolean paramIsNeeded= methodParameterTypes.contains(typeParamName) || methodParameterTypes.contains(typeParamNameAsArray);
+				if (fInstanceUsageRewriter.getTargetMethodhasInstanceUsage() || paramIsNeeded) {
+					//only insert if typeParam not already existing
+					if (!methodTypeParametersNames.contains(typeParameter.getName().getIdentifier())) {
+						typeParamsRewrite.insertLast(typeParameter, null);
+						addNewTypeParamsToJavaDoc(javadoc, typeParameter);
+					}
+				}
+			}
+		}
+		return;
+	}
+
+	private TypeParameter generateTypeParameter(ITypeParameter[] classTypeParameters, int i) throws JavaModelException {
+		String[] bounds= classTypeParameters[i].getBounds();
+		TypeParameter typeParameter= fTargetMethodDeclarationAST.newTypeParameter();
+		typeParameter.setName(fTargetMethodDeclarationAST.newSimpleName(classTypeParameters[i].getElementName()));
+		for (String bound : bounds) {
+			//WildCardTypes are not allowed as bounds
+			fFinalConditionsChecker.checkBoundNotContainingWildCardType(bound);
+			if (!fFinalConditionsChecker.getStatus().hasError()) {
+				SimpleType boundType= fTargetMethodDeclarationAST.newSimpleType(fTargetMethodDeclarationAST.newSimpleName(bound));
+				typeParameter.typeBounds().add(boundType);
+			}
+		}
+		return typeParameter;
+	}
+
+	private List<String> getMethodParameterTypes() {
+		List<SingleVariableDeclaration> methodParams= fTargetMethodDeclaration.parameters();
+		List<String> methodParameterTypes= new ArrayList<>(methodParams.size());
+		for (SingleVariableDeclaration methodParam : methodParams) {
+			Type type= methodParam.getType();
+			//if type is parameterized, then those typeParams need to be included in TypeParamList of method
+			if (type.isParameterizedType()) {
+				ParameterizedType parameterizedType= (ParameterizedType) type;
+				List<Type> typeParamsOfMethodParam= parameterizedType.typeArguments();
+				for (Type typeParamOfMethodParam : typeParamsOfMethodParam) {
+					methodParameterTypes.add(typeParamOfMethodParam.resolveBinding().getName());
```

Comment:
resolveBinding() can return null.  You need to check this here and other places to ensure you don't cause an NPE.